src/provider.c

```diff
@@ -45,6 +46,135 @@ struct p11prov_ctx {
     int nquirks;
 };
 
+static struct p11prov_context_pool {
+    struct p11prov_ctx **contexts;
+    int num;
+
+    pthread_rwlock_t rwlock;
+} ctx_pool = {
+    .contexts = NULL,
+    .num = 0,
+    .rwlock = PTHREAD_RWLOCK_INITIALIZER,
+};
+
+static void fork_prepare(void)
+{
+    int err;
+
+    err = pthread_rwlock_rdlock(&ctx_pool.rwlock);
+    if (err != 0) {
+        err = errno;
+        P11PROV_debug("Can't lock contexts pool (error=%d)", err);
+    }
+
+    for (int i = 0; i < ctx_pool.num; i++) {
+        p11prov_slot_fork_prepare(ctx_pool.contexts[i]->slots);
+    }
+}
+
+static void fork_parent(void)
+{
+    int err;
+
+    for (int i = 0; i < ctx_pool.num; i++) {
+        p11prov_slot_fork_release(ctx_pool.contexts[i]->slots);
+    }
+    err = pthread_rwlock_unlock(&ctx_pool.rwlock);
+    if (err != 0) {
+        err = errno;
+        P11PROV_debug("Failed to release context pool (errno:%d)", err);
+    }
+}
+
+static void fork_child(void)
+{
+    int err;
+
+    /* rwlock, saves TID internally, so we need to reset
+     * after fork in the child */
+    err = pthread_rwlock_init(&ctx_pool.rwlock, NULL);
+    if (err != 0) {
+        err = errno;
+        P11PROV_debug("Failed to reset lock (errno:%d)", err);
```

Comment:
copy&paste error?
```suggestion
        P11PROV_debug("Failed to init lock (errno:%d)", err);
```

Reply:
no this lock was already inited, we are really resetting it (or re-initing if you want), but I think reset is clearer the re-init.